SAP/spartacus
# feat: configurable timeouts for backend API calls

This PR introduces a mechanism of configurable timeout for backend HTTP requests - can be configured different per platform (server/browser). It can be configured via the global config `config.backend.timeout = { browser: ..., server: ...} `. Or it can be configured more specifically - per concrete request by passing a `HTTP_TIMEOUT_CONFIG` `HttpContextToken` to the Angular `HttpClient`'s method.

Context:
Timing out outgoing http calls that take too long is especially important improvement in SSR (Node.js), because in Node.js there is no default timeout of outgoing http calls (in contrary to browsers, which usually timeout long http calls after long time, e.g. 1 minute).

This configurable timeout logic is now implemented in Spartacus, in the layer of Angular Http Interceptor. That said, prior to this feature being released, customers can implement similar logic themselves, e.g. by implementing an Angular Http Interceptor themselves. 

Notes:
- by default the mechanism is disabled, not to introduce a behavioral breaking change in a minor release 5.x

closes #14559
closes https://jira.tools.sap/browse/CXSPA-529

---
**Environments for testing:**
- SSR (`yarn dev:ssr` AND use `curl` in another terminal  e.g.: `curl http://localhost:4200`)
- browser (`yarn start`)

**How can be tested:**
- configure some timeout in the global spartacus config, e.g. (uncomment appropriate:)
    ```ts
    provideConfig({ 
      backend: { 
        timeout: { 
          server: 3_000,
          browser: 3_000
        }
      }
    })
    ```

- Artificially delay the backend responses, e.g. with a custom local proxy:
  - Install `http-proxy` tool
  ```
  npm install -g http-proxy
  ```

  - Write a `http-proxy.js` server file   
  ```js
  // http-proxy.js

  // SPIKE TODO: DON'T MERGE IT

  // PROXY SERVER FOR DELAYING SOME OCC BACKEND RESPONSES
  // ONLY FOR TESTING PURPOSES

  const httpProxy = require('http-proxy');
  const http = require('http');

  const proxy = httpProxy.createProxyServer({ secure: false });

  const ENDPOINT_FOR_DELAY = 'consenttemplates';
  const BACKEND_BASE_URL = 'https://40.76.109.9:9002';
  const DELAY = 3000; // <-------------------------- PLEASE MANIPULATE IT: e.g.  2_500 or 3_000 

  /** custom predicate, whether we should delay a request */
  const shouldDelay = (req) => {
    // Note: In browser there are 2 requests: preflight (OPTIONS) and actual request (GET).

    const result = req.url.includes(ENDPOINT_FOR_DELAY);
    result && console.log({ delay: DELAY, url: req.url, method: req.method });
    return result;
  };

  http
    .createServer(function (req, res) {
      const forwardRequest = () =>
        proxy.web(req, res, { target: BACKEND_BASE_URL });
      const delay = shouldDelay(req) ? DELAY : 0;
      setTimeout(forwardRequest, delay);
    })
    .listen(9002);
  ```

  - Execute the server in a terminal:   
  ```
  node http-proxy.js
  ```
  
- set the OCC backend url to the `http://localhost:9002` (note: NOT HTTPS!) in `.env-cmdrc` file:
```
  "dev": {
    "CX_BASE_URL": "http://localhost:9002"
  },
```
